pkg/router/template/router.go

```diff
@@ -873,69 +873,91 @@ type endpointsCounter func(key string) int32
+	key := endpointsKeyFromParts(route.Namespace, route.Spec.To.Name)
+	info.weight = 1 // default 1 when service weight is omitted
+	info.numEp = counter(key)
+	info.epWeight = 0.0
+	if route.Spec.To.Weight != nil {
+		info.weight = int32(*route.Spec.To.Weight)
 	}
+	if info.numEp > 0 {
+		info.epWeight = float32(info.weight) / float32(info.numEp)
+	}
+	infoData[key] = info
 	for _, svc := range route.Spec.AlternateBackends {
-		if svc.Weight != nil && *svc.Weight > maxWeight {
-			maxWeight = *svc.Weight
+		key = endpointsKeyFromParts(route.Namespace, svc.Name)
+		info.weight = 1
+		info.numEp = counter(key)
+		info.epWeight = 0.0
+		if svc.Weight != nil {
+			info.weight = int32(*svc.Weight)
+		}
+		if info.numEp > 0 {
+			info.epWeight = float32(info.weight) / float32(info.numEp)
+		}
+		infoData[key] = info
+	}
+
+	// maximum specified weight per endpoint
+	var maxEpWeight float32 = 0.0
+	// total specified weight
+	var sumWeight int32 = 0
+
+	for _, info := range infoData {
+		sumWeight += info.weight
+		if info.epWeight > maxEpWeight {
+			maxEpWeight = info.epWeight
 		}
 	}
+
 	// Scale the weights to near the maximum (256).
 	// This improves precision when scaling for the endpoints
-	var scaleWeight int32 = 256 / maxWeight
-
+	var scaleWeight float32 = 0.0
+	if maxEpWeight > 0.0 {
+		scaleWeight = 256.0 / maxEpWeight
+	}
 	// The weight assigned to the service is distributed among the endpoints
 	// for example the if we have two services "A" with weight 20 and 2 endpoints
 	// and "B" with  weight 10 and 4 endpoints the ultimate weights on
 	// endpoints would work out as:
-	// maxWeight = 20, scaleWeight 12 (division truncates 12.8 to 12)
-	// Scaled "A" is 240, "B" is 120
-	// "A" has 2 endpoints: each gets weight 120 (of the available 240)
-	// "B" has 4 endpoints: each gets weight 30  (of the available 120)
-
-	// serviceUnits[key] is the weigth for each endpoint in the service
-	// the sum of the weights of the endpoints is the scaled service weight.
-
-	var numEp int32 = counter(key)
-	if numEp < 1 {
-		numEp = 1
-	}
-	if route.Spec.To.Weight == nil {
-		serviceUnits[key] = scaleWeight / numEp
-	} else {
-		serviceUnits[key] = (*route.Spec.To.Weight * scaleWeight) / numEp
-		if *route.Spec.To.Weight > 0 && serviceUnits[key] < 1 {
+	// service "A" weight per endpoint 10.0
+	// service "B" weight per endpoint 2.5
+	// maximum endpoint weight is 10.0 so scale is 25.6
+	// service "A" scaled endpoint weight 256.0 truncated to 256
+	// service "B" scaled endpoint weight 64.0 truncated to 64
+	// So, all service "A" endpoints get 256 and all service "B" endpoints get 64
+
+	for key, info := range infoData {
+		serviceUnits[key] = int32(info.epWeight * scaleWeight)
+		if info.epWeight > 0.0 && serviceUnits[key] < 1 {
 			serviceUnits[key] = 1
+			glog.V(2).Infof("%s: WARNING: Too many endpoints. Want service load %d, have %d", key, int32(info.epWeight*float32(info.numEp)), info.numEp)
```

Comment:
This is rather scary phrasing.  How about "Warning: The user requested service scale factor for route %s of %d, but only %d could be provided." (and do we need more info to identify what needs to be fiddled?)

And I really wonder if I want to see it in the router logs all the time.  Can we log it once and track it per route?

In order for it to be useful, it needs to end up in the route object itself so the customer can see it.  Spamming the logs with it is probably not useful.

Reply:
Warning: Too many endpoints to achieve desired route weight for %s.  Can have %d endpoints max for balance, have %d.